### ISSUE TITLE
Chrome extension relay: frequent disconnects require manual re-attach (fixable in background.js)

### ISSUE THREAD

**Issue by codexGW** [openclaw/openclaw#15099]
# GitHub Issue Draft: Chrome Extension Relay Disconnects

**Repo:** https://github.com/openclaw/openclaw
**File:** Post as a new issue

---

## Title

Chrome extension relay: frequent disconnects require manual re-attach (fixable in background.js)

## Labels

`bug`, `browser`, `extension`

### Problem

The Chrome extension relay (`background.js`) disconnects frequently during normal use. The user attaches a tab via the toolbar button, uses it successfully, then minutes later the agent gets "tab not found (no attached Chrome tabs for profile 'chrome')." The user must manually re-click the toolbar icon to reattach.

This happens mid-workflow — e.g., agent navigates to site A successfully, then a few minutes later tries to navigate to site B and the connection is gone.

### Environment

- OpenClaw 2026.2.9
- Brave Browser (Chromium-based, supports chrome.debugger)
- macOS 15 (Apple Silicon)
- Extension loaded unpacked via `openclaw browser extension install`

### Root Cause Analysis

After reviewing `background.js`, three issues cause the disconnects:

#### 1. WebSocket drop nukes all debugger sessions (lines ~105-120)

`onRelayClosed()` calls `chrome.debugger.detach()` on **every** attached tab and clears all state when the relay WebSocket drops. Any transient WebSocket hiccup (network blip, service worker restart) destroys all debugger sessions even though the debugger attachment itself was fine.

```js
// Current behavior — relay drop = detach everything
function onRelayClosed(reason) {
  relayWs = null
  // ...
  for (const tabId of tabs.keys()) {
    void chrome.debugger.detach({ tabId }).catch(() => {})  // ← unnecessary
  }
  tabs.clear()
}
```

**Fix:** Don't detach debugger on relay loss. Keep debugger attached, mark relay as disconnected, reconnect WS, then re-announce existing sessions.

#### 2. No WebSocket auto-reconnect

`ws.onclose` and `ws.onerror` only call `onRelayClosed()` — no reconnect attempt. The connection stays dead until the user manually clicks the toolbar icon.

**Fix:** Add exponential backoff reconnect in `onRelayClosed()`:

```js
function onRelayClosed(reason) {
  relayWs = null
  // DON'T detach debugger sessions
  // Schedule reconnect
  scheduleReconnect()
}

let reconnectAttempt = 0
function scheduleReconnect() {
  const delay = Math.min(1000 * 2 ** reconnectAttempt, 30000) + Math.random() * 1000
  reconnectAttempt++
  setTimeout(async () => {
    try {
      await ensureRelayConnection()
      reconnectAttempt = 0
      // Re-announce all still-attached tabs
      for (const [tabId, tab] of tabs.entries()) {
        if (tab.state === 'connected') {
          // Send attachedToTarget events for each tab
        }
      }
    } catch {
      scheduleReconnect()
    }
  }, delay)
}
```

#### 3. MV3 service worker state loss (lines ~19-24)

All tab/session state is in-memory Maps. When Chrome kills the MV3 service worker (~30s idle without activity), state is lost. The debugger may still be attached at the Chrome level, but the extension forgot about it.

Note: Since Chrome 118, active `chrome.debugger` sessions keep the worker alive, and since Chrome 116, WebSocket traffic resets the idle timer. But if both are briefly inactive, the worker can still die.

**Fix:** Persist minimal state (`tabId`, `sessionId`, `targetId`) to `chrome.storage.session`. On worker startup, rehydrate and validate with `chrome.tabs.get()`.

#### 4. No tab lifecycle cleanup (minor)

No `chrome.tabs.onRemoved` or `chrome.tabs.onReplaced` listeners. Closed/replaced tabs leave stale entries in the Maps, which can cause "No attached tab" errors.

**Fix:** Add listeners to clean up state when tabs close or get replaced (prerender swap).

### Impact

This is the biggest friction point when using the Chrome relay. Every disconnect requires physical user intervention (clicking the toolbar button), which breaks automated workflows and makes the relay unreliable for anything beyond single-shot use.

### Suggested Priority

The fixes are straightforward and contained within `background.js`. The most impactful single change is #1 (don't detach on WS drop) combined with #2 (auto-reconnect), which would eliminate most user-facing disconnects.

### References

- [Chrome debugger API - onDetach reasons](https://developer.chrome.com/docs/extensions/reference/api/debugger)
- [MV3 service worker lifecycle](https://developer.chrome.com/docs/extensions/develop/concepts/service-workers/lifecycle)
- [WebSockets in extension service workers](https://developer.chrome.com/docs/extensions/how-to/web-platform/websockets)

**Comment by codexGW** [comment]
## Update: Working Local Patch (v0.1.1)

We built and tested a patched extension that addresses all four root causes from the original report. It has been running stable in Brave (Chromium-based).

### Fixes Implemented

**1. Don't detach debugger on WebSocket drop**
- `onRelayClosed()` no longer calls `chrome.debugger.detach()` on all sessions
- Instead, badges show `…` (connecting) and sessions are preserved

**2. Auto-reconnect WebSocket with exponential backoff**
- Backoff from 1s to 30s with jitter
- On reconnect, re-announces all still-attached tabs to the relay

**3. Persist state to `chrome.storage.session`**
- Tab/session mappings survive MV3 service worker restarts
- Optimistic badge restore on rehydration (set ON immediately, validate async)

**4. Re-attach debugger after cross-origin navigation**
- `onDebuggerDetach` with `target_closed` triggers auto-re-attach with retry (100ms → 300ms → 600ms → 1.2s → 2.5s)
- Skips restricted URLs (`chrome://`, `brave://`, etc.) and retries after they resolve
- `canceled_by_user` reason is respected (user explicitly dismissed debugger)

**5. Badge reliability (MV3 service worker lifecycle)**
- `webNavigation.onCompleted` refreshes badge after page loads
- `tabs.onActivated` refreshes badge when user switches to attached tab
- Keepalive ping every 20s also refreshes badges for all attached tabs
- Added `webNavigation` permission to manifest

**6. Additional cleanup**
- `chrome.tabs.onRemoved` / `onReplaced` listeners for stale entry cleanup
- Debugger listeners registered at module scope (work even when relay is down)
- Pending relay requests timeout after 30s (prevent map leaks)
- `reattachWithRetry` bails if user manually detached during retry window

### Testing Results

- Cross-origin navigation (Amazon → Reddit → YouTube → FAA → Google): badge flickers for <1 second during re-attach, comes back automatically every time
- WebSocket disconnect: sessions preserved, auto-reconnect works
- Service worker restart: state rehydrated from `chrome.storage.session`, badges restored

### Known Limitation

- Brief badge flicker (<1s) during cross-origin navigation is unavoidable — Chrome genuinely detaches the debugger when swapping renderer processes. The re-attach handles it automatically.

### Code

The full patched `background.js` is available if the team wants to reference it for an upstream fix. Happy to open a PR if preferred.

**Comment by graysonpike** [comment]
This sounds like a great improvement. If you could share the fix I'd appreciate it.

**Comment by codexGW** [comment]
Hey — here's the patched `background.js` with all the fixes mentioned in the issue. Been running it for a few days on Brave and haven't needed a single manual re-attach since.

### 1. Don't detach debugger on WebSocket drop
The biggest one. `onRelayClosed()` was calling `chrome.debugger.detach()` on every tab when the relay WS dropped. Now it just marks the relay as disconnected and keeps debugger sessions alive:

```js
function onRelayClosed(reason) {
  relayWs = null
  // Do NOT detach debugger sessions — keep them alive
  // Just update badges to show relay is reconnecting
  for (const [tabId, tab] of tabs.entries()) {
    if (tab.state === 'connected') {
      setBadge(tabId, 'connecting')
    }
  }
  scheduleReconnect()
}
```

### 2. Auto-reconnect with exponential backoff
```js
let reconnectAttempt = 0
function scheduleReconnect() {
  const delay = Math.min(1000 * 2 ** reconnectAttempt, 30000) + Math.random() * 1000
  reconnectAttempt++
  setTimeout(async () => {
    try {
      await ensureRelayConnection()
      reconnectAttempt = 0
      await reannounceAttachedTabs() // re-send attachedToTarget events
    } catch {
      scheduleReconnect()
    }
  }, delay)
}
```

### 3. MV3 service worker state persistence
Persist tab/session state to `chrome.storage.session`. On worker restart, rehydrate and validate (check tab still exists + debugger still attached via `Runtime.evaluate`).

### 4. Re-attach after cross-origin navigation
`onDebuggerDetach` with `reason === 'target_closed'` fires during cross-origin nav (Chrome swaps renderer). Instead of cleaning up, wait briefly and re-attach with retries:

```js
if (reason === 'target_closed') {
  setTimeout(() => reattachWithRetry(tabId, tab, reason, 0), 100)
  return
}
```

### 5. Tab lifecycle cleanup
Added `chrome.tabs.onRemoved` and `chrome.tabs.onReplaced` listeners to clean up stale Map entries.

### 6. Keepalive pings
20s interval ping on the WS to prevent idle service worker termination and reset WS idle timers.

---

Full patched `background.js` is ~600 lines. Happy to put it in a gist or PR if that's useful.

**Comment by derrickburns** [comment]
PR with a fix for this: https://github.com/openclaw/openclaw/pull/15817

**What the patch does (drop-in replacement for `background.js` + adds `alarms` permission to manifest):**

- **Auto-reconnect** — exponential backoff (1s→30s cap, 10 attempts) on WebSocket drop
- **State persistence** — `chrome.storage.local` saves attached tabs/sessions, survives MV3 worker restarts
- **Keepalive alarm** — `chrome.alarms` every 24s (under MV3 30s limit) checks WS health
- **Navigation re-attach** — on `target_closed` debugger detach, waits 500ms then re-attaches
- **Per-tab operation locks** — prevents double-attach race from rapid toolbar clicks
- **Tab lifecycle cleanup** — `onRemoved`/`onUpdated` listeners
- **Request timeouts** — 30s timeout prevents memory leaks
- **Child session cleanup** — proper detach for child sessions

Tested on macOS with Ancestry.com — navigation survival confirmed (previously broke every time).

**Comment by codexGW** [comment]
We've submitted a fix for this in PR #16023. It addresses all the issues mentioned here:

- **Auto-reconnect** with exponential backoff and jitter (unlimited retries)
- **MV3 persistence** via `chrome.storage.session` — survives service worker restarts
- **Seamless cross-origin navigation** — suppresses relay disconnect during `target_closed`, re-attaches after 500ms grace period
- **Stale WebSocket guard** — prevents old socket's close handler from nuking a replacement connection
- **chrome.alarms keepalive** — more reliable than `setInterval` across worker restarts
- **Module-scope debugger listeners** with `whenReady` init gate — events are never missed
- **Tab lifecycle cleanup** — `onRemoved`, `onReplaced` listeners

Tested locally: navigation re-attach and relay reconnect both working smoothly. Also reviewed PR #15817 and adopted its best ideas (operation locks, navigation detach suppression, keepalive catch-path fix) while keeping a different architecture (`chrome.storage.session` over `chrome.storage.local`, unlimited retry, no debugger listener re-registration).
